native/protocol/src/x3dh.rs

```diff
@@ -0,0 +1,364 @@
+use crate::aead::{decrypt_data, encrypt_data, AeadError};
+use crate::bundle::*;
+use chacha20poly1305::{
+    aead::{KeyInit, Payload},
+    ChaCha20Poly1305,
+};
+use ed25519_dalek::{Signature, SigningKey, VerifyingKey};
+use hkdf::Hkdf;
+use serde::{Deserialize, Serialize};
+use sha2::Sha256;
+use thiserror::Error;
+use x25519_dalek::{
+    PublicKey as X25519PublicKey, ReusableSecret as X25519ReusableSecret,
+    StaticSecret as X25519StaticSecret,
+};
+
+// See https://signal.org/docs/specifications/x3dh/ for an explanation of the below variable names
+// and functions of the X3DH protocol.
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SignedPreKey {
+    pub pre_key: X25519PublicKey,
+    pub signature: Signature,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SignedPreKeys {
+    pub pre_keys: Vec<X25519PublicKey>,
+    pub signature: Signature,
+}
+
+pub struct X3DHSendKeyAgreement {
+    pub ek: X25519PublicKey,
+    pub sk: [u8; 32],
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Message {
+    pub sender_identity: String,
+    pub sender_ik: VerifyingKey,
+    pub ek: X25519PublicKey,
+    pub opk: Option<X25519PublicKey>,
+    pub ciphertext: Vec<u8>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PreKeyBundle {
+    pub ik: VerifyingKey,
+    pub opk: Option<X25519PublicKey>,
+    pub spk: SignedPreKey,
+}
+
+// KDF(KM) represents 32 bytes of output from the HKDF algorithm [3] with inputs:
```

Comment:
I deeply appreciate the explanatory comments, would love you long time if you included the expanded form next to the acronym.